File: frontend/src/lib/stores/mounts.ts
// The mounts store holds the active mounts and drives mount/unmount (§7.11.6).
import { writable } from "svelte/store";

import { ListMounts, MountFs, UnmountFs } from "../../../wailsjs/go/app/App";
import type { app } from "../../../wailsjs/go/models";

export const mounts = writable<app.MountDTO[]>([]);
export const mountsError = writable<app.ErrorDTO | null>(null);

export async function refreshMounts(): Promise<void> {
  try {
    const res = await ListMounts();
    if (res.error) {
      mountsError.set(res.error);
    } else {
      mounts.set(res.mounts ?? []);
      mountsError.set(null);
    }
  } catch {
    // binding unavailable outside the webview
  }
}

export async function doMount(fs: string, mountPoint: string, mountType: string): Promise<app.ErrorDTO | null> {
  const err = await MountFs(fs, mountPoint, mountType);
  await refreshMounts();
  return err ?? null;
}

export async function doUnmount(mountPoint: string): Promise<app.ErrorDTO | null> {
  const err = await UnmountFs(mountPoint);
  await refreshMounts();
  return err ?? null;
}

File: app/mounts.go
package app

import (
	"context"
	"time"

	"github.com/conductor-app/conductor/internal/core/domain"
)

// MountDTO is one active mount for the frontend.
type MountDTO struct {
	Fs         string `json:"fs"`
	MountPoint string `json:"mountPoint"`
	MountedOn  string `json:"mountedOn"`
}

// MountsResultDTO is the list of active mounts or a typed error.
type MountsResultDTO struct {
	Mounts []MountDTO `json:"mounts"`
	Error  *ErrorDTO  `json:"error"`
}

// ListMounts returns the active mounts (§7.11.6).
func (a *App) ListMounts() MountsResultDTO {
	ms, err := a.mounts.List(context.Background())
	if err != nil {
		return MountsResultDTO{Error: errorToDTO(err)}
	}
	out := make([]MountDTO, 0, len(ms))
	for _, m := range ms {
		out = append(out, toMountDTO(m))
	}
	return MountsResultDTO{Mounts: out}
}

// MountFs mounts fs at mountPoint. mountType may be "".
func (a *App) MountFs(fs, mountPoint, mountType string) *ErrorDTO {
	if err := a.mounts.Mount(context.Background(), fs, mountPoint, mountType); err != nil {
		return errorToDTO(err)
	}
	return nil
}

// UnmountFs unmounts the mount at mountPoint.
func (a *App) UnmountFs(mountPoint string) *ErrorDTO {
	if err := a.mounts.Unmount(context.Background(), mountPoint); err != nil {
		return errorToDTO(err)
	}
	return nil
}

func toMountDTO(m domain.Mount) MountDTO {
	dto := MountDTO{Fs: m.Fs, MountPoint: m.MountPoint}
	if !m.MountedOn.IsZero() {
		dto.MountedOn = m.MountedOn.Format(time.RFC3339)
	}
	return dto
}
